[Merged by Bors] - work parallel worker queue work

antonlerner opened the pull request:
## Motivation
try to run the queue in parallel to not deadlock itself
## Changes
<!-- Please describe in detail the changes made -->

## Test Plan
<!-- Please specify how these changes were tested 
(e.g. unit tests, manual testing, etc.) -->

## TODO
<!-- This section should be removed when all items are complete -->
- [ ] Explain motivation or link existing issue(s)
- [ ] Test changes and document test plan
- [ ] Update documentation as needed

AmitShaul committed:
WIP, added elk yaml files

sudachen committed:
elastic search k8s fix

sudachen committed:
storageclass fix

AmitShaul committed:
WIP, added elk conf files, added fixture to run elk, need to remove former elastic handler

AmitShaul committed:
Merge branch 'develop' of github.com:spacemeshos/go-spacemesh into elk

AmitShaul committed:
WIP ES context to work with new elk, fixed a bug in kibana deployment text file

AmitShaul committed:
WIP added logstass, added debugging code that should be removed

sudachen committed:
Update elastic_stateful.yaml

sudachen committed:
Update service.yaml

Do not publish elastic to the public Internet

sudachen committed:
Merge pull request #2154 from spacemeshos/sudachen-patch-1

Update elastic_stateful.yaml

sudachen committed:
some fixes for ELK

AmitShaul committed:
WIP elk works with kubernetes metadata, still sleeping in mining test

AmitShaul committed:
elk integrated and working, should add dumping of elastic data to a central elastic on teardown

AmitShaul committed:
Merge branch 'develop' of github.com:spacemeshos/go-spacemesh into elk

AmitShaul committed:
fixed a bug when printing unsuccessful miners in query_message

sudachen committed:
logstash and elastic must use SSD storage not HDD

AmitShaul committed:
new CI is working, changed logstash stateful

AmitShaul committed:
Merge branch 'develop' of github.com:spacemeshos/go-spacemesh into elk

antonlerner committed:
trying a fix

AmitShaul committed:
WIP all tests but hares tests are passing, commit for testing with Antons fix

AmitShaul committed:
Merge branch 'sync_atxs' of github.com:spacemeshos/go-spacemesh into elk_hare

AmitShaul committed:
WIP using elasticdump to dump ES data, not all issues resolved with es_dump

AmitShaul committed:
added support for ES dumping using ES API

AmitShaul committed:
deprication with extensions/v1beta1, changed deployment to work with app/v1 instead, changed k8s automation config files to support the newer version

AmitShaul committed:
changed ES main URL

AmitShaul committed:
fixed oracle k8s object with older and deprecated extensions/v1beta1

AmitShaul committed:
Merge branch 'develop' of github.com:spacemeshos/go-spacemesh into elk_hare

AmitShaul committed:
fixed namespace bug in filebeat, NAMESPACE was switched with a static namespace. handled exceptions in es_reindex(). fixed NAMESPACE replace regex with negative look behind for _ char.

AmitShaul committed:
added timeout for es_dump + removed v1beta api when deleting deployment

AmitShaul committed:
merged support for fluent bit, added sleep time before starting p2p tests

AmitShaul committed:
Merge branch 'develop' of github.com:spacemeshos/go-spacemesh into elk_hare

AmitShaul committed:
fixed logstash input bug, fixed es_dump bug

AmitShaul committed:
[*] namespace support for fluent bit, now fluent bit will only collect logs from its own namespace.
[*] kube deployment files with modified namespaces deletion happens right after file was loaded to memory.
[*] added tolerance to elastic_logging k8s files.
[*] added pod affinity to fluent bit daemonset
[*] added namespace labels to k8s client, bootstrapoet and poet files (only for those with no ss)

AmitShaul committed:
merged dev_elk
[*] wait for elk to be ready
[*] changed k8s resources request and limit to be equal
[*] fixed main ES toleration
[*] fixed Fluent bit to collect only sm files as input
[*] replaced podAffinity with nodeSelector in fluent bit
[*] added support for namespacing when applying statefulset k8s file

AmitShaul committed:
added support for node pool per test
need to fix delayed nodes

AmitShaul committed:
fixed makefile and ci.yml to work with the new flow

AmitShaul committed:
ci.yml fixed ci-stage2 needs to be the new tests

AmitShaul committed:
ci.yml fixed ci-stage2 status

AmitShaul committed:
Makefiles dockerbuild-test-elk, fixed bug with dockerbuild-test-elk - changed DockerFileTestsElk to DockerFileTests

AmitShaul committed:
fixed genesis timing
[*] lowered genesis time for add\remove nodes, late nodes and hare tests
[*] fixed number of nodes when deploying node pool in NodePoolDep

AmitShaul committed:
fixed bugs
[*] test mining raised genesis delta to 50
[*] test_p2p raised pods to 60

AmitShaul committed:
[*] added support for ES reinstallation incase of an exception
[*] added handler for k8s files applyment
[*] adjust genesis delta in add\remove nodes and test_bs

AmitShaul committed:
[*] changed add_curl fixture not to depand on setup_bootstrap
[*] changed add_curl to happen before deploying clients and bootstrap
[*] raised create_pod timeout to 70
[*] raised remove_nodes genesis_delta to 70

AmitShaul committed:
DEBUGGING added sleeping time after ES and logstash timeout exception for debugging purposes

AmitShaul committed:
removed debugging from wait_for_minimal_elk_cluster_ready, raised es timeout to 240 and ls timeout to 180

web3rover committed:
run on custom runner

web3rover committed:
changed runner name

AmitShaul committed:
removed es removal incase of an exception, raised logstash timeout to 240, after fixing github actions timeout should be no more than 120

AmitShaul committed:
Merge branch 'elk_hare' of github.com:spacemeshos/go-spacemesh into elk_hare

web3rover committed:
use ubuntu-latest runner

web3rover committed:
Merge branch 'elk_hare' of github.com:spacemeshos/go-spacemesh into elk_hare

AmitShaul committed:
fixed 500 exception by adding retries when posting a curl request to a client pod

AmitShaul committed:
fixed utils api_call function

AmitShaul committed:
[*] set a new ES service
[*] changed the order of elk and pool fixtures in order to allow node pool to be deleted before ES dumping is happening

antonlerner committed:
work parallel worker queue work

antonlerner committed:
fmt

antonlerner committed:
start syncer after p2p has peers

antonlerner committed:
added test adjustments to first atx created in third epoch

AmitShaul committed:
raised hare test genesis time from 40 to 55

y0sher committed:
Linting fixes

y0sher committed:
implement new interface for mocks

antonlerner committed:
after fixing vet

nkryuchkov committed:
Fix tests in #2239 (#2240)

* Fix unit tests and make some code improvements

# Conflicts:
#	activation/poetlistener_test.go
#	sync/syncer.go

* Fix genesis voting test

* Fix imports

antonlerner commented:
bors try

AmitShaul committed:
[*] new and stronger ES, changed the ES main IP under config.py
[*] added a new argument when running pytest tests, is-dump, when set to true logs will be dumped to the main ES server when tests are finished.
[*] removed old commands from Makefile and added is-dump argument when running new pytest elk tests

AmitShaul committed:
[*] changed is-dump to be a boolean
[*] added keep-alive header to ES dump request

AmitShaul committed:
changed is-dump to be boolean

antonlerner commented:
bors merge

antonlerner commented:
bors try

AmitShaul committed:
removed keep-alive header from ES dump

noamnelke commented on sync/fetch_queue.go:
```diff
@@ -103,9 +112,9 @@ func (fq *fetchQueue) addToPending(ids []types.Hash32) []chan bool {
 	for _, id := range ids {
 		ch := make(chan bool, 1)
 		deps = append(deps, ch)
-		if _, ok := fq.pending[id]; !ok {
-			idsToAdd = append(idsToAdd, id)
-		}
+		//if _, ok := fq.pending[id]; !ok {
+		idsToAdd = append(idsToAdd, id)
+		//}
```
Not sure why the conditional was removed here...

It seems that without the conditional `idsToAdd` is an exact copy of `ids`.

If this change was accidental, let's revert it. Otherwise, let's remove the commented out code and also the `idsToAdd` variable and use `ids` instead.

noamnelke commented on sync/fetch_queue.go:
```diff
@@ -67,27 +67,36 @@ func (fq *fetchQueue) work() error {
-			return nil
-		}
+	wg := sync.WaitGroup{}
+	wg.Add(runtime.NumCPU())
+	for i := 0; i < runtime.NumCPU(); i++ {
```
`fq.handleFetch(bjb)` is the actual work happening in these threads? I assume it's mostly CPU-bound and that's why you use `runtime.NumCPU()` as the concurrency factor?

In any case, I'd make a constant/variable for this instead of calling `runtime.NumCPU()` twice. Would make the code clearer IMO and be slightly less error prone.

noamnelke reviewed (approved):
LGTM, only minor comments.

nkryuchkov committed:
Fix test comments

antonlerner committed:
Merge branch 'sync_work_worker' of https://github.com/spacemeshos/go-spacemesh into sync_work_worker

antonlerner committed:
fixed start stop panic and CR comments

antonlerner committed:
fixed merge issue

antonlerner committed:
start stop deadlock

AmitShaul committed:
removed is-dump option, only failed tests will dump logs into main ES server when tests are done

antonlerner committed:
fixed sync start tesdt

y0sher commented:
bors try

antonlerner commented:
bors merge

AmitShaul committed:
[*] polling for ES documents status after dumping instead of waiting for response from _reindex
[*] fixed execution function to print to screen only with conditions to prevent console flooding.

antonlerner commented:
bors merge

AmitShaul committed:
added is_dump option, when set to true (t, yes, 1..) logs will be dumped to  main ES even if tests have passed

y0sher commented:
bors merge

AmitShaul committed:
PR #2234 fixes, added some timing prints and other very minor fixes

AmitShaul committed:
[*] removed ES credentials from code, we now relay that the user will have both ES_USER and ES_PASS environment variables set before running automation code.
[*] duplicate_file_and_replace_phrases to handle a list of phrases to change and a list of values to replace with.

AmitShaul committed:
added ES_USER and ES_PASS to makefile and github actions file

antonlerner committed:
Merge remote-tracking branch 'origin/elk_hare' into sync_work_worker

y0sher commented:
bors merge
